### PR TITLE
Increase HTTP client timeout in catalog2dnx

### PR DESCRIPTION
**Pull request by joelverhagen**
Address https://github.com/NuGet/Engineering/issues/1261.

- Add `-httpClientTimeoutInSeconds` option to catalog2dnx.
- Only download the .nupkg once
- Download the .nupkg to disk and use this stream for getting .nuspec and for uploading to flat container
- Improve the tests a lot
- Only accept the HTTP response if it is 200 OK. We previously allowed any 2XX.

### CODE REVIEW

**Comment by joelverhagen** [comment]
Oh, I set the HTTP client timeout to 900 seconds. This is because the download of a 250 megabyte file takes approximately 7 minutes on our jobs machine. I doubled-ish this value to 15 minutes.

File: src/Catalog/PackageDownloader.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+namespace NuGet.Services.Metadata.Catalog
+{
+    public class PackageDownloader
+    {
+        private const int BufferSize = 8192;
+
+        private readonly HttpClient _httpClient;
+        private readonly ILogger _logger;
+
+        public PackageDownloader(HttpClient httpClient, ILogger logger)
```
Comment (joelverhagen) [inline src/Catalog/PackageDownloader.cs]: This is a copy of `PackageDownloader` from NuGet.Jobs. I have removed some custom telemetry events and returned `null` on 404.

File: src/Catalog/PackageDownloader.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                    packageStream.Length,
+                    stopwatch.Elapsed.TotalSeconds,
+                    packageUri);
+
+                return packageStream;
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(
+                    0,
```
Comment (shishirx34) [inline src/Catalog/PackageDownloader.cs]: parameter name

File: src/Catalog/Dnx/DnxCatalogCollector.cs [hunk 3]
```diff
@@ -36,67 +45,67 @@ public DnxCatalogCollector(Uri index, StorageFactory storageFactory, ITelemetryS
...
+
+            var packageDownloader = new PackageDownloader(client, _logger);
+            var requestUri = Utilities.GetNugetCacheBustingUri(sourceUri);
+
+            using (var stream = await packageDownloader.DownloadAsync(requestUri, cancellationToken))
             {
-                httpResponseMessage.EnsureSuccessStatusCode();
+                if (stream == null)
+                {
+                    Trace.TraceInformation("package not found...");
```
Comment (loic-sharma) [inline src/Catalog/Dnx/DnxCatalogCollector.cs]: >TraceInformation [](start = 26, length = 16)

Should this be a warning?

File: src/Catalog/PackageDownloader.cs (hunk 2 above)
Comment (joelverhagen) [reply src/Catalog/PackageDownloader.cs]: I'll do `(EventId)0`. This seems like a good middle ground to know what that `0` means and not having to name all further parameters.

File: src/Catalog/Dnx/DnxCatalogCollector.cs (hunk 3 above)
Comment (joelverhagen) [reply src/Catalog/Dnx/DnxCatalogCollector.cs]: Wasn't before, but I can switch this.

**Pull request merged**